### PR TITLE
Add functions and gRPC server

### PR DESCRIPTION
**Pull request by FelixKuhnAnsys**
- added config
- added dockerfile
- populated externalfunctions (including slight changes to the functions; i.e. removal of session context)
- added functiondefinitions to fill internalstates available functions on launch
- added grpc protos file and generate grpcdefinition
- added grpc server to expose three function: listFunctions, runFunction, streamFunction
- added internalstates

### CODE REVIEW

File: pkg/grpcdefinition/externalfunctions.pb.go
Comment (RobPasMue) [inline pkg/grpcdefinition/externalfunctions.pb.go:1]: These files should not be committed. They should be generated as part of the build process.

File: pkg/grpcdefinition/externalfunctions_grpc.pb.go
Comment (RobPasMue) [inline pkg/grpcdefinition/externalfunctions_grpc.pb.go:1]: These files should not be committed. They should be generated as part of the build process.

File: pkg/protos/externalfunctions.proto
Comment (RobPasMue) [inline pkg/protos/externalfunctions.proto:1]: I would recommend to add protofile docstrings here

File: pkg/protos/externalfunctions.proto [hunk 1]
```diff
@@ -0,0 +1,69 @@
+syntax = "proto3";
+
+package externalfunctions;
+
+option go_package = "./grpcdefinition";
+
+// Exposed Service
+service ExternalFunctions {
+    rpc ListFunctions(ListFunctionsRequest) returns (ListOfFunctions) {}
```
Comment (RobPasMue) [inline pkg/protos/externalfunctions.proto:9]:
```suggestion
    rpc ListFunctions(ListFunctionsRequest) returns (ListFunctionsResponse) {}
```

File: pkg/protos/externalfunctions.proto [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    rpc RunFunction(FunctionInputs) returns (FunctionOutputs) {}
+    rpc StreamFunction(FunctionInputs) returns (stream StreamOutput) {}
+}
+
+// ListFunctions
+message ListFunctionsRequest {
+}
+
+message ListOfFunctions {
```
Comment (RobPasMue) [inline pkg/protos/externalfunctions.proto:18]:
```suggestion
message ListFunctionsResponse {
```

**Review by RobPasMue — CHANGES_REQUESTED** [review]
I've got some concerns here, mostly with the proto files. We shouldn't be committing the autogenerated code. We should be generating it on the fly when building the library.

Also, it seems like the docstrings are not being build into the docs. We need to modify the docs/api section to include the new sections

File: pkg/grpcdefinition/externalfunctions.pb.go
Comment (FelixKuhnAnsys) [reply pkg/grpcdefinition/externalfunctions.pb.go:1]: as discussed with andreas for now we leave it and change this at a later point

File: pkg/grpcdefinition/externalfunctions_grpc.pb.go
Comment (FelixKuhnAnsys) [reply pkg/grpcdefinition/externalfunctions_grpc.pb.go:1]: as discussed with andreas for now we leave it and change this at a later point

**Comment by FelixKuhnAnsys** [comment]
> I've got some concerns here, mostly with the proto files. We shouldn't be committing the autogenerated code. We should be generating it on the fly when building the library.
> 
> Also, it seems like the docstrings are not being build into the docs. We need to modify the docs/api section to include the new sections

maybe we can move the whole grpc definition to sharedtypes at a later point? otherwise we would need another public repo for it. for now andreas said to keep it here do to time constraints.

not sure where/how to modify the docs/api, maybe we can do it together on monday and you can show me? i think this can be done in a seperate pr though

File: pkg/protos/externalfunctions.proto (hunk 1 above)
Comment (FelixKuhnAnsys) [reply pkg/protos/externalfunctions.proto]: done

File: pkg/protos/externalfunctions.proto
Comment (FelixKuhnAnsys) [reply pkg/protos/externalfunctions.proto:1]: done

File: pkg/protos/externalfunctions.proto (hunk 2 above)
Comment (FelixKuhnAnsys) [reply pkg/protos/externalfunctions.proto]: done

**Review by RobPasMue — APPROVED** [review]

**Pull request merged**